nodes/student_localiser.py

```diff
@@ -0,0 +1,660 @@
+			os.makedirs(self.save_dir+'norm/')
+		if not os.path.isdir(self.save_dir+'pose/'):
+			os.makedirs(self.save_dir+'pose/')
+		if not os.path.isdir(self.save_dir+'offset/'):
+			os.makedirs(self.save_dir+'offset/')
+		if not os.path.isdir(self.save_dir+'correction/'):
+			os.makedirs(self.save_dir+'correction/')
+		if self.save_full_res_images or self.save_full_res_images_at_goal:
+			if not os.path.isdir(self.save_dir+'full/'):
+				os.makedirs(self.save_dir+'full/')
+		self.goal_number = 0
+
+		self.save_params()
+
+
+	def setup_publishers(self):
+		self.goal_pub = rospy.Publisher('goal', Goal, queue_size=1)
+		if self.publish_gt_goals:
+			self.goal_pose_pub = rospy.Publisher('goal_pose', PoseStamped, queue_size=1)
+		rospy.wait_for_service('match_image')
+		self.match_image = rospy.ServiceProxy('match_image', ImageMatch, persistent=True)
+
+	def setup_subscribers(self):
+		if not self.ready:
+			self.srv_ready = rospy.Service('ready_localiser', Trigger, self.on_ready)
+		self.sub_odom = rospy.Subscriber("odom", Odometry, self.process_odom_data, queue_size=1)
+		self.sub_images = rospy.Subscriber('image', Image, self.process_image_data, queue_size=1, buff_size=2**22)
+		self.tfBuffer = tf2_ros.Buffer()
+		self.tfListener = tf2_ros.TransformListener(self.tfBuffer)
+
+		# For Multi-Robot System
+		self.sub_teacher_pose = rospy.Subscriber("teacher_pose", Pose, self.append_teacher_pose, queue_size=1)
+
+	def append_teacher_pose(self, msg):
+		'''Append new teacher pose to the student list when ever it arrived'''
+		self.poses.append(msg)
+
+	def save_params(self):
+		params = {
+			'load_dir' : self.load_dir,
+			'rotation_correction_gain' : self.rotation_correction_gain,
+			'path_correction_gain' : self.path_correction_gain,
+			'wait_for_ready' : not self.ready,
+			'global_localisation_init' : self.global_localisation_init,
+			'resize': self.resize,
+			'patch_size': self.patch_size,
+			'stop_at_end': self.stop_at_end,
+			'discrete_correction': self.discrete_correction,
+			'goal_distance_spacing': GOAL_DISTANCE_SPACING,
+			'lookahead_distance_ratio': LOOKAHEAD_DISTANCE_RATIO,
+			'turning_target_theta_range_distance_ratio': TURNING_TARGET_RANGE_DISTANCE_RATIO,
+			'goal_theta_tolerance': GOAL_THETA_TOLERANCE,
+			'image_recognition_threshold': self.image_recognition_threshold,
+			'image_subsampling': CORR_SUB_SAMPLING,
+			'image_field_of_view_width_deg': FIELD_OF_VIEW_DEG,
+			'search_range': self.search_range,
+			'save_full_res_images': self.save_full_res_images,
+			'save_gt_data': self.save_gt_data
+		}
+		with open(self.save_dir + 'params.txt', 'w') as params_file:
+			params_file.write(json.dumps(params))
+
+	def on_ready(self, srv):
+		if not self.ready:
+			self.start()
+			self.ready = True
+			return TriggerResponse(success=True)
+		else:
+			return TriggerResponse(success=False, message="Localiser already started.")
+
+	def start(self):
+		while (len(self.poses) < self.frames_delay):
+			pass
```

Comment:
Like in the other file it gets stuck in this while loop until the teacher has published enough poses?

Reply:
Yes